pkg/requests/builder.go

```diff
@@ -0,0 +1,173 @@
+package requests
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/bitly/go-simplejson"
+)
+
+// Builder allows users to construct a request and then either get the requests
+// response via Do(), parse the response into a simplejson.Json via JSON(),
+// or to parse the json response into an object via Into().
+type Builder interface {
+	WithContext(context.Context) Builder
+	WithBody(io.Reader) Builder
+	WithMethod(string) Builder
+	WithHeader(http.Header) Builder
+	SetHeader(key, value string) Builder
+	Do() (*http.Response, error)
+	Into(interface{}) error
+	JSON() (*simplejson.Json, error)
+}
+
+type builder struct {
+	context  context.Context
+	method   string
+	endpoint string
+	body     io.Reader
+	header   http.Header
+	response *http.Response
+}
+
+// New provides a new Builder for the given endpoint.
+func New(endpoint string) Builder {
+	return &builder{
+		endpoint: endpoint,
+		method:   "GET",
+	}
+}
+
+// WithContext adds a context to the request.
+// If no context is provided, context.Background() is used instead.
+func (r *builder) WithContext(ctx context.Context) Builder {
+	r.context = ctx
+	return r
+}
+
+// WithBody adds a body to the request.
+func (r *builder) WithBody(body io.Reader) Builder {
+	r.body = body
+	return r
+}
+
+// WithMethod sets the request method. Defaults to "GET".
+func (r *builder) WithMethod(method string) Builder {
+	r.method = method
+	return r
+}
+
+// WithHeader replaces the request header map with the given header map.
+func (r *builder) WithHeader(header http.Header) Builder {
+	r.header = header
+	return r
+}
+
+// SetHeader sets a single header to the given value.
+// May be used to add multiple headers.
+func (r *builder) SetHeader(key, value string) Builder {
+	if r.header == nil {
+		r.header = make(http.Header)
+	}
+	r.header.Set(key, value)
+	return r
+}
+
+// Do performs the request and returns the response in its raw form.
+// If the request has already been performed, returns the previous result.
+// This will not allow you to repeat a request.
+func (r *builder) Do() (*http.Response, error) {
+	if r.response != nil {
+		// Request has already been done
+		return r.response, nil
+	}
+
+	// Must provide a non-nil context to NewRequestWithContext
+	if r.context == nil {
+		r.context = context.Background()
+	}
+
+	req, err := http.NewRequestWithContext(r.context, r.method, r.endpoint, r.body)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+	req.Header = r.header
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error performing request: %v", err)
+	}
+
+	r.response = resp
+	return resp, nil
+}
+
+// Into performs the request and attempts to unmarshal the response into the
+// the given interface. The response body is assumed to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) Into(into interface{}) error {
```

Comment:
Should the Into/JSON method variants be aligned to the `Unmarshal` naming scheme that encoding libraries use (and that you are doing under the hood)?

Reply:
I wonder if there's a clean way to register the target interface to Unmarshal into (or multiples...?) as a builder modifier in the chain (so the chain finishes with the `Do/Send` verb).  Error handling might get hairy trying to do it that way though.  (I'm just thinking out loud).

Reply:
We could change this up to have `Do` return a `Result`, that result could then be reused a few times if you wanted to `Into` a couple of things, but I haven't really seen any use case for that in our code at present.

`UnmarshalJSON` and `UnmarshalInto` make sense though

Reply:
I came across this dated project that tried to do HTTP request via builder pattern.  They seem to have implemented better than I could explain about having the unmarshal destination being a builder modifier method before the final execution:

https://github.com/BrandonRomano/wrecker

https://github.com/BrandonRomano/wrecker/blob/d37dcbf1a77802e496cddbcc55e28e5ee9583459/wrecker.go#L118

Could eliminate the need for the standalone sermi-duplicated `Into` that operated on a response body in the cases where you needed to evaluate the response & unmarshal into a struct -- would result in always getting both back if an `UnmarshalInto` or something was tagged on the builder.

Not sure how this fits with the `simplejson` output case though.

Reply:
I don't know how I feel about this. I don't think we could implement the simplejson method doing it that way so then we would have two patterns. What you you think of Do returning a `Result` which is a struct containing the response and a possible error, that could then have methods `UnmarshalInto` and `UnmarshalJSON` on them? I think I'd prefer that over the wrecker approach personally.

I think there were only two cases where we go the response and then unmarshalled into an object, I'm trying to think what this approach would save us, 1 error check?

Reply:
I like that idea a lot, seems like a good way to simplify the chaining while supporting simplejson as an output option.

The multiple executors per output format felt weird to me.  This route have a consistent executor `Do` that can be chained into a transformer `UnmarshalInto` and `UnmarshalJSON` feels very clean.

Reply:
What about access to the original error when `Do()` is chained off of by the transformer? Should `UnmarshalInto` and `UnmarshalJSON` return the original `Result` + unmarshal error so that the `Result.Error` field is accessible after the transformation attempt?  Or `UnmarshalInto` / `UnmarshalJSON` error immediately when `Result.Error` is non-Nil and only return `error` like normal unmarshal methods?

Chaining is so much harder in Golang with error outputs.  Much easier in other languages when the middle parts just throw a bazillion potential exceptions that developers frequently forget to handle correctly 😄

Reply:
Great, I'll have a go at that.

And I'd expect the unmarshal methods to look pretty much identical to now, except they'd check for an error in the result and return that early if it's there